orb-ui/cone/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+pub mod button;
+pub mod lcd;
+pub mod led;
+
+use crate::button::Button;
+use crate::lcd::{Lcd, LcdCommand};
+use crate::led::{Led, CONE_LED_COUNT};
+use color_eyre::eyre;
+use color_eyre::eyre::Context;
+use embedded_graphics::pixelcolor::{Rgb565, RgbColor};
+use ftdi_embedded_hal::libftd2xx::{Ft4232h, Ftdi, FtdiCommon};
+use image::{ImageFormat, Luma};
+use orb_rgb::Argb;
+use std::sync::{Arc, Mutex};
+use std::{env, fs};
+use tokio::sync::mpsc;
+
+const CONE_FTDI_DEVICE_COUNT: usize = 8;
+
+#[derive(Debug)]
+enum Status {
+    Connected,
+    Disconnected,
+}
+
+/// Cone can be created only if connected to the host over USB.
+pub struct Cone {
+    connection_status: Arc<Mutex<Status>>,
+    lcd: Lcd,
+    led_strip: Arc<Mutex<Led>>,
+    _button: Button,
+}
+
+pub enum ConeEvents {
+    ButtonPressed(bool),
+}
+
+impl Cone {
+    /// Create a new Cone instance.
+    pub fn new(event_queue: mpsc::UnboundedSender<ConeEvents>) -> eyre::Result<Self> {
+        let connection_status = if ftdi_embedded_hal::libftd2xx::list_devices()?.len()
+            != CONE_FTDI_DEVICE_COUNT
+        {
+            return Err(eyre::eyre!(
+                "FTDI device count mismatch: cone not connected?"
+            ));
+        } else {
+            let mut device: Ft4232h = Ftdi::with_index(6)?
+                .try_into()
+                .wrap_err("Failed to initialize FTDI device")?;
+            device.reset().wrap_err("Failed to reset")?;
+            Arc::new(Mutex::new(Status::Connected))
+        };
+
+        let lcd = Lcd::spawn()?;
+        let led_strip = Led::spawn()?;
+        let _button = Button::spawn(event_queue.clone(), connection_status.clone())?;
+
+        let cone = Cone {
+            connection_status,
+            lcd,
+            led_strip,
+            _button,
+        };
+
+        Ok(cone)
+    }
+
+    pub fn is_connected(&self) -> bool {
+        if let Ok(status) = self.connection_status.lock() {
+            matches!(*status, Status::Connected)
+        } else {
+            false
+        }
+    }
+
+    /// Update the RGB LEDs by passing the values to the LED strip sender.
+    pub fn queue_rgb_leds(
+        &mut self,
+        pixels: &[Argb; CONE_LED_COUNT],
+    ) -> eyre::Result<()> {
+        self.led_strip
+            .lock()
+            .expect("cannot lock LED strip mutex")
+            .clone_tx()
```

Comment:
If you don't use an Arc<Mutex<Led>>, the &mut self will allow you to avoid the lock and avoid the clone_tx.

Reply:
done